Assets/MixedRealityToolkit.Tests/PlayModeTests/InteractableTests.cs

```diff
@@ -61,6 +61,146 @@ public IEnumerator TestAddInteractableAtRuntime()
             yield return null;
         }
 
+        /// <summary>
+        /// Tests that an interactable component can be added to a GameObject
+        /// at runtime, and an OnClick event handler can be added
+        /// </summary>
+        /// <returns></returns>
+        [UnityTest]
+        public IEnumerator TestAddOnClickRuntime()
+        {
+            var cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
+            cube.transform.position = Vector3.forward;
+            PlayModeTestUtilities.PushHandSimulationProfile();
+            PlayModeTestUtilities.SetHandSimulationMode(HandSimulationMode.Gestures);
+
+            // This should not throw an exception
+            var interactable = cube.AddComponent<Interactable>();
+
+            // Subscribe to interactable's on click so we know the click went through
+            bool wasClicked = false;
+            interactable.OnClick.AddListener(() => { wasClicked = true; });
+
+            TestHand th = new TestHand(Handedness.Right);
+            yield return th.Show(Vector3.zero);
+
+            CameraCache.Main.transform.LookAt(interactable.transform);
+            yield return th.SetGesture(ArticulatedHandPose.GestureId.Pinch);
+            yield return th.SetGesture(ArticulatedHandPose.GestureId.Open);
+
+            Assert.True(wasClicked);
+
+            // clean up
+            GameObject.Destroy(cube);
+            PlayModeTestUtilities.PopHandSimulationProfile();
+
+            yield return null;
+        }
+
+        /// <summary>
+        /// Tests that an interactable component can be added to a GameObject
+        /// at runtime, and all receivers that extend ReceiverBase
+        /// and have event handlers can easily be added, and work
+        /// </summary>
+        /// <returns></returns>
+        [UnityTest]
+        public IEnumerator TestRuntimeEvents()
+        {
+            var cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
+            cube.transform.position = Vector3.right;
+
+            var interactable = cube.AddComponent<Interactable>();
+
+            // OnFocus
+            var onFocusReceiver = interactable.AddReceiver<InteractableOnFocusReceiver>();
+
+            bool didHover = false;
+            bool didUnHover = false;
+            onFocusReceiver.OnFocusOn.AddListener(() => didHover = true);
+            onFocusReceiver.OnFocusOff.AddListener(() => didUnHover = true);
+            CameraCache.Main.transform.LookAt(interactable.transform);
+            yield return null;
+            CameraCache.Main.transform.LookAt(Vector3.forward);
+            yield return null;
+            Assert.True(didHover, "Interactable did not receive hover event");
+            Assert.True(didUnHover, "Interactable did not receive un-hover event");
+
+            // OnGrab
```

Comment:
Move these all out into methods, I think.

Reply:
The test method is a bit long, but it's within the scope of the naming of this function, and the code is readable within this test as well, so it's fine IMO.

The other thing to consider is having each thing in its own test (i.e. TestOnGrab, TestOnFocus)... etc, that way if only one thing fails, the other tests at least get to run, and you get more granular information about what failed.

---
In reply to: [320324131](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5854#discussion_r320324131) [](ancestors = 320324131)

Reply:
Actually yeah, the second suggestion is probably better, but it's up to you.

---
In reply to: [320468834](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5854#discussion_r320468834) [](ancestors = 320468834,320324131)